core/src/main/scala/kafka/log/Log.scala

```diff
@@ -84,13 +84,19 @@ case class LogAppendInfo(var firstOffset: Option[Long],
                          targetCodec: CompressionCodec,
                          shallowCount: Int,
                          validBytes: Int,
-                         offsetsMonotonic: Boolean) {
+                         offsetsMonotonic: Boolean,
+                         lastOffsetOfFirstBatch: Long) {
   /**
    * Get the first offset if it exists, else get the last offset.
    * @return The offset of first message if it exists; else offset of the last message.
    */
   def firstOrLastOffset: Long = firstOffset.getOrElse(lastOffset)
 
+  /**
+   * Get the first offset if it exists, else get the last offset of the first batch
+   */
+  def firstOrLastOfFirstBatchOffset: Long = firstOffset.getOrElse(lastOffsetOfFirstBatch)
```

Comment:
The only other place that uses firstOrLastOffset is for maybeCacheTxnFirstOffsetMetadata(). I am wondering if that can/should also use firstOrLastOfFirstBatchOffset. Perhaps @hachikuji can comment on this.

Reply:
For transactions, we should be able to rely on having the first offset available, so using `firstOrLastOfFirstBatchOffset` should be fine. We could even use `firstOffset`, but that may require some reworking of the append logic.

By the way, I wonder if a friendlier name for this method might be `firstOffsetEstimate` or something like that?